Fix issues found by mypy checks

PR #127 for issue #57 added checkmypy build task and CI job. That job found issues that need to be resolved. 

```
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:4: error: Skipping analyzing "pyqir.parser._native": module is installed, but missing library stubs or py.typed marker
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:4: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:607: error: Incompatible types in assignment (expression has type "QirType", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:608: error: Incompatible return value type (got "None", expected "QirType")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:940: error: Incompatible types in assignment (expression has type "List[QirInstr]", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:941: error: Incompatible return value type (got "None", expected "List[QirInstr]")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:950: error: Incompatible types in assignment (expression has type "QirTerminator", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:951: error: Incompatible return value type (got "None", expected "QirTerminator")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:962: error: Incompatible types in assignment (expression has type "List[QirPhiInstr]", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:963: error: Incompatible return value type (got "None", expected "List[QirPhiInstr]")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:998: error: Incompatible types in assignment (expression has type "QirType", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:999: error: Incompatible return value type (got "None", expected "QirType")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:1028: error: Incompatible types in assignment (expression has type "List[QirParameter]", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:1029: error: Incompatible return value type (got "None", expected "List[QirParameter]")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:1037: error: Incompatible types in assignment (expression has type "QirType", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:1038: error: Incompatible return value type (got "None", expected "QirType")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:1046: error: Incompatible types in assignment (expression has type "List[QirBlock]", variable has type "None")
/home/runner/work/pyqir/pyqir/pyqir-parser/pyqir/parser/_parser.py:1047: error: Incompatible return value type (got "None", expected "List[QirBlock]")
/home/runner/work/pyqir/pyqir/pyqir-evaluator/pyqir/evaluator/_nonadaptiveevaluator.py:5: error: Skipping analyzing "pyqir.evaluator._native": module is installed, but missing library stubs or py.typed marker
Found 18 errors in 2 files (checked 10 source files)
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -145,20 +145,3 @@ jobs:
           branch: gh-pages
           folder: docs/_build/html
         if: ${{ github.event_name == 'push' }}
-  mypy:
-    runs-on: "ubuntu-20.04"
-    needs: build
-    steps:
-      - uses: actions/checkout@v2
-        with:
-          submodules: false
-
-      - name: Download Build Artifacts
-        uses: actions/download-artifact@v2
-        with:
-          name: manylinux-artifacts
-          path: target/wheels
-
-      - name: Check type annotations and mypy stub files
-        run: ./build.ps1 -t checkmypy
-        shell: pwsh

--- a/eng/psakefile.ps1
+++ b/eng/psakefile.ps1
@@ -60,7 +60,7 @@ task manylinux -depends build-manylinux-container-image, run-manylinux-container
 
 task musllinux -depends build-musllinux-container-image, run-musllinux-container-image, run-examples-in-musl-containers
 
-task checks -depends cargo-fmt, cargo-clippy
+task checks -depends cargo-fmt, cargo-clippy, checkmypy
 
 task rebuild -depends qirlib, generator, evaluator, parser
 

--- a/pyqir-evaluator/pyqir/evaluator/_nonadaptiveevaluator.py
+++ b/pyqir-evaluator/pyqir/evaluator/_nonadaptiveevaluator.py
@@ -2,7 +2,7 @@
 # Licensed under the MIT License.
 
 from pyqir.evaluator._gateset import GateSet
-from pyqir.evaluator._native import PyNonadaptiveJit
+from pyqir.evaluator._native import PyNonadaptiveJit # type: ignore
 from typing import List, Optional
 
 

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -603,13 +603,9 @@ class QirInstr:
         Gets the QirType instance representing the output of this instruction. If the instruction
         has no output, the type will be an instance of QirVoidType.
         """
-        if self._type is None:
-            ty = QirType(self.instr.type)
-            self._type = ty
-        else:
-            ty = self._type
-        return ty
-
+        if self._instructions is None:
+            self._instructions = [QirInstr(i) for i in self.block.instructions]
+        return cast(List[QirInstr], self._instructions)
 
 class QirOpInstr(QirInstr):
     """
@@ -949,7 +945,7 @@ class QirBlock:
         Gets the terminator instruction for this block. Every block has exactly one terminator
         and it is the last intruction in the block.
         """
-        if self._terminator == None:
+        if self._terminator is None:
             self._terminator = QirTerminator(self.block.terminator)
         return cast(QirTerminator, self._terminator)
 
@@ -961,7 +957,7 @@ class QirBlock:
         of phi nodes, but they are always the first instructions in any given block. A block with no
         phi nodes will return an empty list.
         """
-        if self._phi_nodes == None:
+        if self._phi_nodes is None:
             self._phi_nodes = [QirPhiInstr(i) for i in self.block.phi_nodes]
         return cast(List[QirPhiInstr], self._phi_nodes)
 
@@ -997,7 +993,7 @@ class QirParameter:
         """
         Gets the type of this parameter as represented in the QIR.
         """
-        if self._type == None:
+        if self._type is None:
             self._type = QirType(self.param.type)
         return cast(QirType, self._type)
 
@@ -1026,7 +1022,7 @@ class QirFunction:
         """
         Gets the list of parameters used when calling this function.
         """
-        if self._parameters == None:
+        if self._parameters is None:
             self._parameters = [QirParameter(i) for i in self.func.parameters]
         return cast(List[QirParameter], self._parameters)
 
@@ -1035,7 +1031,7 @@ class QirFunction:
         """
         Gets the return type for this function.
         """
-        if self._return_type == None:
+        if self._return_type is None:
             self._return_type = QirType(self.func.return_type)
         return cast(QirType, self._return_type)
 
@@ -1044,7 +1040,7 @@ class QirFunction:
         """
         Gets all the basic blocks for this function.
         """
-        if self._blocks == None:
+        if self._blocks is None:
             self._blocks = [QirBlock(i) for i in self.func.blocks]
         return cast(List[QirBlock], self._blocks)
 
@@ -1120,7 +1116,7 @@ class QirModule:
         """
         Gets all the functions defined in this module.
         """
-        if self._functions == None:
+        if self._functions is None:
             self._functions = [QirFunction(i) for i in self.module.functions]
         return self._functions
 
@@ -1147,7 +1143,7 @@ class QirModule:
         """
         Gets any functions with the "EntryPoint" attribute.
         """
-        if self._entrypoint_funcs == None:
+        if self._entrypoint_funcs is None:
             self._entrypoint_funcs = [
                 QirFunction(i) for i in self.module.get_entrypoint_funcs()
             ]
@@ -1158,7 +1154,7 @@ class QirModule:
         """
         Gets any functions with the "InteropFriendly" attribute.
         """
-        if self._interop_funcs == None:
+        if self._interop_funcs is None:
             self._interop_funcs = [
                 QirFunction(i) for i in self.module.get_interop_funcs()
             ]

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -11,8 +11,8 @@ from pyqir.parser._native import (
     PyQirOperand,
     PyQirType,
     module_from_bitcode
-)
-from typing import List, Optional, Tuple
+) # type: ignore
+from typing import cast, List, Optional, Tuple
 
 __all__ = [
     "QirType",
@@ -605,7 +605,7 @@ class QirInstr:
         """
         if self._type == None:
             self._type = QirType(self.instr.type)
-        return self._type
+        return cast(QirType, self._type)
 
 
 class QirOpInstr(QirInstr):
@@ -938,7 +938,7 @@ class QirBlock:
         """
         if self._instructions == None:
             self._instructions = [QirInstr(i) for i in self.block.instructions]
-        return self._instructions
+        return cast(List[QirInstr], self._instructions)
 
     @property
     def terminator(self) -> QirTerminator:
@@ -948,7 +948,7 @@ class QirBlock:
         """
         if self._terminator == None:
             self._terminator = QirTerminator(self.block.terminator)
-        return self._terminator
+        return cast(QirTerminator, self._terminator)
 
     @property
     def phi_nodes(self) -> List[QirPhiInstr]:
@@ -960,7 +960,7 @@ class QirBlock:
         """
         if self._phi_nodes == None:
             self._phi_nodes = [QirPhiInstr(i) for i in self.block.phi_nodes]
-        return self._phi_nodes
+        return cast(List[QirPhiInstr], self._phi_nodes)
 
     def get_phi_pairs_by_source_name(self, name: str) -> List[Tuple[str, QirOperand]]:
         """
@@ -996,7 +996,7 @@ class QirParameter:
         """
         if self._type == None:
             self._type = QirType(self.param.type)
-        return self._type
+        return cast(QirType, self._type)
 
 
 class QirFunction:
@@ -1025,7 +1025,7 @@ class QirFunction:
         """
         if self._parameters == None:
             self._parameters = [QirParameter(i) for i in self.func.parameters]
-        return self._parameters
+        return cast(List[QirParameter], self._parameters)
 
     @property
     def return_type(self) -> QirType:
@@ -1034,7 +1034,7 @@ class QirFunction:
         """
         if self._return_type == None:
             self._return_type = QirType(self.func.return_type)
-        return self._return_type
+        return cast(QirType, self._return_type)
 
     @property
     def blocks(self) -> List[QirBlock]:
@@ -1043,7 +1043,7 @@ class QirFunction:
         """
         if self._blocks == None:
             self._blocks = [QirBlock(i) for i in self.func.blocks]
-        return self._blocks
+        return cast(List[QirBlock], self._blocks)
 
     @property
     def required_qubits(self) -> Optional[int]:

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -587,7 +587,7 @@ class QirInstr:
 
     def __init__(self, instr: PyQirInstruction):
         self.instr = instr
-        self._type = None
+        self._type: Optional[QirType] = None
 
     @property
     def output_name(self) -> Optional[str]:
@@ -917,9 +917,9 @@ class QirBlock:
 
     def __init__(self, block: PyQirBasicBlock):
         self.block = block
-        self._instructions = None
-        self._terminator = None
-        self._phi_nodes = None
+        self._instructions: Optional[List[QirInstr]] = None
+        self._terminator: Optional[QirTerminator] = None
+        self._phi_nodes: Optional[List[QirPhiInstr]] = None
 
     @property
     def name(self) -> str:
@@ -979,7 +979,7 @@ class QirParameter:
 
     def __init__(self, param: PyQirParameter):
         self.param = param
-        self._type = None
+        self._type: Optional[QirType] = None
 
     @property
     def name(self) -> str:
@@ -1007,10 +1007,9 @@ class QirFunction:
 
     def __init__(self, func: PyQirFunction):
         self.func = func
-        self._parameters = None
-        self._parameters = None
-        self._return_type = None
-        self._blocks = None
+        self._parameters: Optional[List[QirParameter]] = None
+        self._return_type: Optional[QirType] = None
+        self._blocks: Optional[List[QirBlock]] = None
 
     @property
     def name(self) -> str:
@@ -1042,8 +1041,6 @@ class QirFunction:
         """
         Gets all the basic blocks for this function.
         """
-        self._blocks: Optional[List[QirBlock]] # Required to avoid type checking error
-
         if self._blocks == None:
             self._blocks = [QirBlock(i) for i in self.func.blocks]
         return self._blocks

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -1,7 +1,7 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT License.
 
-from pyqir.parser._native import (
+from pyqir.parser._native import ( # type: ignore
     PyQirModule,
     PyQirFunction,
     PyQirParameter,
@@ -11,7 +11,7 @@ from pyqir.parser._native import (
     PyQirOperand,
     PyQirType,
     module_from_bitcode
-) # type: ignore
+)
 from typing import cast, List, Optional, Tuple
 
 __all__ = [

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -1042,7 +1042,7 @@ class QirFunction:
         """
         Gets all the basic blocks for this function.
         """
-        self._blocks: Optional[List[QirBlock]] # Required for type hinting
+        self._blocks: Optional[List[QirBlock]] # Required to avoid type checking error
 
         if self._blocks == None:
             self._blocks = [QirBlock(i) for i in self.func.blocks]

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -603,9 +603,12 @@ class QirInstr:
         Gets the QirType instance representing the output of this instruction. If the instruction
         has no output, the type will be an instance of QirVoidType.
         """
-        if self._type == None:
-            self._type = QirType(self.instr.type)
-        return cast(QirType, self._type)
+        if self._type is None:
+            ty = QirType(self.instr.type)
+            self._type = ty
+        else:
+            ty = self._type
+        return ty
 
 
 class QirOpInstr(QirInstr):

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -1042,6 +1042,8 @@ class QirFunction:
         """
         Gets all the basic blocks for this function.
         """
+        self._blocks: Optional[List[QirBlock]] # Required for type hinting
+
         if self._blocks == None:
             self._blocks = [QirBlock(i) for i in self.func.blocks]
         return self._blocks

--- a/pyqir-parser/pyqir/parser/_parser.py
+++ b/pyqir-parser/pyqir/parser/_parser.py
@@ -603,9 +603,10 @@ class QirInstr:
         Gets the QirType instance representing the output of this instruction. If the instruction
         has no output, the type will be an instance of QirVoidType.
         """
-        if self._instructions is None:
-            self._instructions = [QirInstr(i) for i in self.block.instructions]
-        return cast(List[QirInstr], self._instructions)
+        if self._type == None:
+            self._type = QirType(self.instr.type)
+        return cast(QirType, self._type)
+
 
 class QirOpInstr(QirInstr):
     """
@@ -945,7 +946,7 @@ class QirBlock:
         Gets the terminator instruction for this block. Every block has exactly one terminator
         and it is the last intruction in the block.
         """
-        if self._terminator is None:
+        if self._terminator == None:
             self._terminator = QirTerminator(self.block.terminator)
         return cast(QirTerminator, self._terminator)
 
@@ -957,7 +958,7 @@ class QirBlock:
         of phi nodes, but they are always the first instructions in any given block. A block with no
         phi nodes will return an empty list.
         """
-        if self._phi_nodes is None:
+        if self._phi_nodes == None:
             self._phi_nodes = [QirPhiInstr(i) for i in self.block.phi_nodes]
         return cast(List[QirPhiInstr], self._phi_nodes)
 
@@ -993,7 +994,7 @@ class QirParameter:
         """
         Gets the type of this parameter as represented in the QIR.
         """
-        if self._type is None:
+        if self._type == None:
             self._type = QirType(self.param.type)
         return cast(QirType, self._type)
 
@@ -1022,7 +1023,7 @@ class QirFunction:
         """
         Gets the list of parameters used when calling this function.
         """
-        if self._parameters is None:
+        if self._parameters == None:
             self._parameters = [QirParameter(i) for i in self.func.parameters]
         return cast(List[QirParameter], self._parameters)
 
@@ -1031,7 +1032,7 @@ class QirFunction:
         """
         Gets the return type for this function.
         """
-        if self._return_type is None:
+        if self._return_type == None:
             self._return_type = QirType(self.func.return_type)
         return cast(QirType, self._return_type)
 
@@ -1040,7 +1041,7 @@ class QirFunction:
         """
         Gets all the basic blocks for this function.
         """
-        if self._blocks is None:
+        if self._blocks == None:
             self._blocks = [QirBlock(i) for i in self.func.blocks]
         return cast(List[QirBlock], self._blocks)
 
@@ -1116,7 +1117,7 @@ class QirModule:
         """
         Gets all the functions defined in this module.
         """
-        if self._functions is None:
+        if self._functions == None:
             self._functions = [QirFunction(i) for i in self.module.functions]
         return self._functions
 
@@ -1143,7 +1144,7 @@ class QirModule:
         """
         Gets any functions with the "EntryPoint" attribute.
         """
-        if self._entrypoint_funcs is None:
+        if self._entrypoint_funcs == None:
             self._entrypoint_funcs = [
                 QirFunction(i) for i in self.module.get_entrypoint_funcs()
             ]
@@ -1154,7 +1155,7 @@ class QirModule:
         """
         Gets any functions with the "InteropFriendly" attribute.
         """
-        if self._interop_funcs is None:
+        if self._interop_funcs == None:
             self._interop_funcs = [
                 QirFunction(i) for i in self.module.get_interop_funcs()
             ]

